homeassistant/components/notify/legacy.py

```diff
@@ -101,26 +101,27 @@ async def async_setup_platform(
             )
             hass.config.components.add(f"{DOMAIN}.{integration_name}")
 
-    setup_tasks = [
-        asyncio.create_task(async_setup_platform(integration_name, p_config))
-        for integration_name, p_config in config_per_platform(config, DOMAIN)
-        if integration_name is not None
-    ]
-
-    if setup_tasks:
-        await asyncio.wait(setup_tasks)
-
-    async def async_platform_discovered(
-        platform: str, info: DiscoveryInfoType | None
-    ) -> None:
-        """Handle for discovered platform."""
-        await async_setup_platform(platform, discovery_info=info)
-
-    hass.data[NOTIFY_DISCOVERY] = (
-        discovery.async_listen_platform(hass, DOMAIN, async_platform_discovered)
-        if hass.data[NOTIFY_DISCOVERY]
-        else None
-    )
+    async with hass.data[NOTIFY_DISCOVERY]["lock"]:
```

Comment:
I wonder if we can move the synchronization to the reload helper so we don't need to guard inside every integration that wants to use the reload helper.

Reply:
We coudl, but we still need to store the async.Lock().

Reply:
Storing the lock, or whatever we'll use for synchronization, in hass.data is not a problem. It would be nice if we don't need to modify the integrations that use the reload helper to accomplish the synchronization.